src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+import java.util.TreeSet;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.ibm.icu.impl.Pair;
+import com.yubico.webauthn.AssertionRequest;
+import com.yubico.webauthn.AssertionResult;
+import com.yubico.webauthn.FinishAssertionOptions;
+import com.yubico.webauthn.FinishRegistrationOptions;
+import com.yubico.webauthn.RegistrationResult;
+import com.yubico.webauthn.RelyingParty;
+import com.yubico.webauthn.StartAssertionOptions;
+import com.yubico.webauthn.StartRegistrationOptions;
+import com.yubico.webauthn.data.AuthenticatorAssertionResponse;
+import com.yubico.webauthn.data.AuthenticatorAttestationResponse;
+import com.yubico.webauthn.data.AuthenticatorSelectionCriteria;
+import com.yubico.webauthn.data.ByteArray;
+import com.yubico.webauthn.data.ClientAssertionExtensionOutputs;
+import com.yubico.webauthn.data.ClientRegistrationExtensionOutputs;
+import com.yubico.webauthn.data.PublicKeyCredential;
+import com.yubico.webauthn.data.PublicKeyCredentialCreationOptions;
+import com.yubico.webauthn.data.ResidentKeyRequirement;
+import com.yubico.webauthn.data.UserIdentity;
+import com.yubico.webauthn.data.UserVerificationRequirement;
+
+import it.smartcommunitylab.aac.SystemKeys;
+import it.smartcommunitylab.aac.core.service.SubjectService;
+import it.smartcommunitylab.aac.model.Subject;
+import it.smartcommunitylab.aac.webauthn.model.WebAuthnCredentialCreationInfo;
+import it.smartcommunitylab.aac.webauthn.persistence.WebAuthnCredential;
+import it.smartcommunitylab.aac.webauthn.persistence.WebAuthnCredentialsRepository;
+import it.smartcommunitylab.aac.webauthn.persistence.WebAuthnUserAccount;
+import it.smartcommunitylab.aac.webauthn.persistence.WebAuthnUserAccountRepository;
+
+public class WebAuthnRpService {
+
+    WebAuthnUserAccountRepository webAuthnUserAccountRepository;
+
+    WebAuthnCredentialsRepository webAuthnCredentialsRepository;
+
+    SubjectService subjectService;
+
+    final private RelyingParty rp;
+    final String provider;
+
+    private static Long TIMEOUT = 9000L;
+
+    // TODO: civts make it so this gets cleaned from time to time
+    private Map<String, WebAuthnCredentialCreationInfo> activeRegistrations = new ConcurrentHashMap<>();
+    private Map<String, AssertionRequest> activeAuthentications = new ConcurrentHashMap<>();
+
+    public WebAuthnRpService(RelyingParty rp,
+            WebAuthnUserAccountRepository webAuthnUserAccountRepository,
+            WebAuthnCredentialsRepository webAuthnCredentialsRepository,
+            SubjectService subjectService,
+            String provider) {
+        this.rp = rp;
+        this.provider = provider;
+        this.webAuthnCredentialsRepository = webAuthnCredentialsRepository;
+        this.webAuthnUserAccountRepository = webAuthnUserAccountRepository;
+        this.subjectService = subjectService;
+    }
+
+    public Pair<PublicKeyCredentialCreationOptions, String> startRegistration(String username, String realm,
```

Comment:
this method requires refactoring, too many operations and the return object is not well defined.

Reply:
Redefined return object to existing DTO, `WebAuthnRegistrationResponse`, in f5dda8a98fb396dccffe3f21455aa6b931bbd3f6